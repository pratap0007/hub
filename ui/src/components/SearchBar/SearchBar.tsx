import { TextInput } from '@patternfly/react-core';
import React, { useState } from 'react';

import './SearchBar.css';

const SearchBar: React.FC = () => {
  const [value, setValue] = useState('');

  const setInput = (event: string) => {
    setValue(event);
  };

  return (
    <TextInput
      value={value}
      type="search"
      onChange={setInput}
      aria-label="text input example"
      placeholder="Search for resources..."
      spellCheck="false"
      className="search-bar"
    />
  );
};

export default SearchBar;
