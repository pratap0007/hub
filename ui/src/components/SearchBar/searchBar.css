.pf-c-search-input input {
  width: 20em;
}
