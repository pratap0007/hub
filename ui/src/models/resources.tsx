import { types, flow, getEnv, Instance, getSnapshot } from 'mobx-state-tree';


const Catalog = types.model("Catalog", {
    id: types.optional(types.number, 0),
    type: types.optional(types.string, ""),
})

export const Tag = types.model("Tag", {
    id: types.optional(types.number, 0),
    name: types.optional(types.string, " ")


})

export const LatestVersion = types.model("Version", {
    id: types.number,
    version: types.optional(types.string, "0.1"),
    displayName: types.string,
    description: types.string,
    minPipelinesVersion: types.optional(types.string, ""),
    rawURL: types.string,
    webURL: types.string,
    updatedAt: types.string,
})

export const ResourceItem = types.model("ResourceItem", {
    id: types.number,
    type: types.string,
    name: types.string,
    catalog: Catalog,
    latestVersion: LatestVersion,
    tags: types.optional(types.array(Tag), []),
    rating: types.optional(types.number, 0),

})

export const Resources = types.model("Resources", {
    resourceList: types.optional(types.array(ResourceItem), []),
    isLoading: types.optional(types.boolean, false)
})
    .views((self) => ({
        get count() {
            return self.resourceList.length;
        },
        get fetch() {
            return getEnv(self).fetch || window.fetch.bind(self);
        },
    }))
    .actions((self) => ({
        add(item: IResourceItem) {
            self.resourceList.push(item)

        }

    }))
    .actions((self) => ({
        load: flow(function* () {
            console.log("7777", { self })

            try {
                const resp = yield window.fetch('/resources.json').then((res: any) => res.json())
                resp.forEach((item: IResourceItem) => { self.add(item) })
                console.log(getSnapshot(self.resourceList))
            } catch (error) {
                console.error("Failed to fetch resources", error)
            }

        })
    }))
    .actions((self) => ({
        afterCreate() {
            self.load();
        },

    }))

export type IResourceItem = Instance<typeof ResourceItem>