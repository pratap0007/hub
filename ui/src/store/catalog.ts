import { types, Instance } from "mobx-state-tree";

export const Catalog = types
  .model({
    name: types.string,
    selected: false
  })
  .actions(self => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICatalog = Instance<typeof Catalog>;

export const CatalogStore = types
  .model("CatalogStore", {
    catalogList: types.array(Catalog)
  })

  .views(self => ({
    get count() {
      return self.catalogList.length;
    },
    get list() {
      const { catalogList } = self;
      return catalogList;
    },
    get selectedCatalogType() {
      const { catalogList } = self;
      return catalogList
        .filter((type: ICatalog) => type.selected)
        .map((type: ICatalog) => type.name);
    }
  }))

  .actions(self => ({
    add(item: any) {
      self.catalogList.push(item);
    },
    toggleCatalogType(kindName: string) {
      self.catalogList.forEach((kind: ICatalog) => {
        if (kind.name === kindName) {
          kind.toggle();
          return;
        }
      });
    },
    clearAll() {
      self.catalogList.map((item: ICatalog) => (item.selected = false));
    }
  }));
