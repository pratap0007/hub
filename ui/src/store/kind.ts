import { types, Instance } from "mobx-state-tree";

export const Kind = types
  .model({
    name: types.string,
    selected: false
  })
  .actions(self => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type IKind = Instance<typeof Kind>;

export const KindStore = types
  .model({
    kindList: types.array(Kind)
  })

  .views(self => ({
    get count() {
      return self.kindList.length;
    },
    get selectedkind() {
      const { kindList } = self;
      return kindList
        .filter((type: IKind) => type.selected)
        .map((type: IKind) => type.name);
    },
    get list() {
      const { kindList } = self;
      return kindList;
    }
  }))

  .actions(self => ({
    add(item: any) {
      self.kindList.push(item);
    },
    toggleKind(kindName: string) {
      self.kindList.forEach((kind: IKind) => {
        if (kind.name === kindName) {
          kind.toggle();
          return;
        }
      });
    },
    clearAll() {
      self.kindList.map((item: IKind) => (item.selected = false));
    }
  }));

export type IKindStore = Instance<typeof KindStore>;
