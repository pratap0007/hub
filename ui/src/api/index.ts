import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource, IVersion } from '../store/resource';

interface Token {
  token: string;
  refreshInterval: string;
  expiresAt: number;
}

interface TokenData {
  access: Token;
  refresh: Token;
}

export interface AuthResponse {
  data: TokenData;
}

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
  resourceVersion(resourceId: number): Promise<IVersion>;
  versionUpdate(versionId: number): Promise<IVersion>;
  authentication(authCode: string): Promise<AuthResponse>;
}

export class Hub implements Api {
  async categories() {
    try {
      return axios.get(`${API_URL}/categories`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resources() {
    try {
      return axios.get(`${API_URL}/resources`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async authentication(authCode: string) {
    try {
      return axios.post(`${API_URL}/auth/login?code=${authCode}`).then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async resourceVersion(resourceId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/${resourceId}/versions`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }

  async versionUpdate(versionId: number) {
    try {
      return axios
        .get(`${API_URL}/resource/version/${versionId}`)
        .then((response) => response.data);
    } catch (err) {
      return err.response;
    }
  }
}
