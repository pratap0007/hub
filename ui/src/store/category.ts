import { AxiosError } from 'axios';
import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Category = types
  .model('Category', {
    id: types.identifierNumber,
    name: types.string,
    selected: false
  })

  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICategory = Instance<typeof Category>;
export type ICategoryStore = Instance<typeof CategoryStore>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.map(Category),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.size;
    },

    get values() {
      return Array.from(self.items.values());
    },

    // This returns list of selected category's name
    get selectedByName() {
      return Array.from(self.items.values())
        .filter((c: ICategory) => c.selected)
        .reduce((acc: string[], c: ICategory) => [...acc, c.name], []);
    },

    // This returns set of selected category
    get selected(): Set<number> {
      const list: Set<number> = new Set();
      self.items.forEach((c: ICategory) => {
        if (c.selected) {
          list.add(c.id);
        }
      });
      return list;
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.put(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clearSelected() {
      self.items.forEach((c) => {
        c.selected = false;
      });
    },

    toggleByName(name: string) {
      self.items.forEach((c) => {
        if (c.name === name) {
          c.selected = true;
        }
      });
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;

        const json = yield api.categories();

        // creating the model only after the store has the tags normalized
        const categories: ICategory[] = json.data.map((c: ICategory) => ({
          id: c.id,
          name: c.name
        }));

        categories.forEach((c: ICategory) => self.add(c));
      } catch (error) {
        const err = error as AxiosError;
        self.err = err.toString();
      }

      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
