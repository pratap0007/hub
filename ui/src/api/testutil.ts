import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
