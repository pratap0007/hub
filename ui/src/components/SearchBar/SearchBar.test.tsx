import React from 'react';
import renderer from 'react-test-renderer';
import SearchBar from './SearchBar';
import { IResourceStore, ResourceStore } from '../../store/resources';
import { when } from 'mobx';
import { shallow, mount } from 'enzyme';
import { CatalogStore } from '../../store/catalog';
import { CategoryStore } from '../../store/category';
import { KindStore } from '../../store/kind';

import { FakeHub } from '../../api/testutil';
const TESTDATA_DIR = `src/store/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('searchBar', () => {
  it('should render componnet correctly', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );
    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        const component = shallow(<SearchBar store={store} />);
        expect(component).toMatchSnapshot();

        done();
      }
    );
  });

  it('set text input and trigger event', (done) => {
    const onSearchChange = jest.fn();
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );
    when(
      () => !store.isLoading,
      () => {
        const wrapper = shallow(<SearchBar store={store} />);
        const event = {
          preventDefault() {},
          target: {
            value: 'This is just for test'
          }
        };

        const onSearchChange = jest.fn();

        wrapper.find('.search-bar').simulate('change', { target: { value: '2018-01-04' } });
        expect(onSearchChange).toHaveBeenCalledTimes(1);
        done();
      }
    );
  });
});
