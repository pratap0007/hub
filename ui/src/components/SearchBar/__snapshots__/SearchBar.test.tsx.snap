// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<input
  aria-invalid={false}
  aria-label="text input example"
  className="pf-c-form-control search-bar"
  disabled={false}
  onBlur={[Function]}
  onChange={[Function]}
  onFocus={[Function]}
  placeholder="Search for resources..."
  readOnly={false}
  required={false}
  spellCheck="false"
  type="search"
  value=""
/>
`;
